src/accel/HepMC3PrimaryGenerator.cc

```diff
@@ -212,10 +192,73 @@ void HepMC3PrimaryGenerator::GeneratePrimaryVertex(G4Event* g4_event)
     }
 
     CELER_VALIDATE(g4_event->GetNumberOfPrimaryVertex() > 0,
-                   << "event " << evt.event_number()
+                   << "event " << g4_event->GetEventID()
                    << " did not contain any primaries suitable for "
                       "simulation");
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Read the given event from the file in a thread-safe manner.
+ *
+ * Each event can only be read once. Because reading across threads may be out
+ * of order, the next event to read may not be the next event in the file. To
+ * fix this with minimal performance and memory impact, we read all events up
+ * to the one requested into a buffer. Once the events are buffered, we release
+ * the shared pointer (marking its location in the buffer as empty) and return
+ * it to the calling thread. Before reading new events, empty elements at the
+ * front of the buffer are released. In the usual case, the buffer should only
+ * be size(num_threads), but in the worst case (the first event is very slow
+ * and the other threads keep processing new events) it can be arbitrarily
+ * large. However, since accessing an element in a deque is a constant-time
+ * operation, this function should be constant time at best and scale with the
+ * number of threads at worst.
+ */
+auto HepMC3PrimaryGenerator::read_event(size_type event_id) -> SPHepEvt
+{
+    CELER_EXPECT(event_id < num_events_);
+
+    std::lock_guard scoped_lock{read_mutex_};
+    CELER_EXPECT(event_id >= start_event_);
+
+    // Remove empties at the front of the deque
+    while (!event_buffer_.empty() && !event_buffer_.front())
+    {
+        event_buffer_.pop_front();
+        ++start_event_;
+    }
+
+    CELER_LOG_LOCAL(debug) << "Reading to event " << event_id
+                           << ": buffer has [" << start_event_ << ", "
+                           << start_event_ + event_buffer_.size() << ")";
+
+    // Read new events until we get to the requested one
+    while (event_id >= start_event_ + event_buffer_.size())
+    {
+        size_type expected_id = start_event_ + event_buffer_.size();
+        event_buffer_.push_back(std::make_shared<HepMC3::GenEvent>());
+        reader_->read_event(*event_buffer_.back());
+
+        auto read_evt_id = event_buffer_.back()->event_number();
+        CELER_VALIDATE(!reader_->failed(),
+                       << "event " << expected_id << " could not be read");
+
+        if (static_cast<size_type>(read_evt_id) != expected_id)
+        {
+            CELER_LOG(warning)
+                << "HepMC3 event IDs are not consecutive from zero: ID "
+                << read_evt_id << " is not expected ID " << expected_id;
```

Comment:
This seems like a reasonable assumption, but are we sure this is the case (and does it matter)? I haven't e.g. seen anything from HepMC3 suggesting the event IDs couldn't just be arbitrary (and some of the event files we use, like the one for our celer-g4 test, start at 1).

Reply:
Geant4 starts at zero, so if we don't have that event do we just die immediately on the worker thread? Or read until the end of the file? What if the file has multiple events with the same ID? Bad planning on hepmc's part...

Reply:
Oh is the problem that we need to preserve the original IDs from the event record because they'll be needed at other stages in the simulation chain? I guess I'm missing why it matters if the events get assigned new IDs by Geant4. What do we do if some event records start at 0 and others at 1?

Reply:
That's why here were basically pretending the event ids on the file are sequential from zero. Before we pretended that *and* gave the event to the first core that asked for it.

Reply:
In Geant4, the event ID is assigned by the number of events processed and it may not be necessarily same to one assigned by an external generator (or an event ID record, if any, in an input event file) - so it does not matter whether the EventID starts in 0 or 1, or nothing in input.  Only important thing may be the order of event-reading  or processing the sequence of events by the Geant4 event manager for `reproducibility`, especially for the MT mode - there are two different cases: one for the full reproducibility regardless of the number threads used (i.e, seeds are assigned in the run-level), another one for the reproducibility when the same number of threads is used (i.e., seeds are assigned in the thread-level).

Reply:
Thanks, yeah that makes sense. Since we're warning if the HepMC3 ID doesn't match the Geant4 ID I was wondering if there were any reason we might need to preserve the former, but it sounds like no (or at least like it would be trivial to create a mapping now that the ID sequence matches the order in the event file).